Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -294,6 +294,140 @@ namespace Babylon
             texture->Width = width;
             texture->Height = height;
         }
+
+        class IndexBufferData final
+        {
+        public:
+            IndexBufferData(Napi::TypedArray bytes, uint16_t flags, bool dynamic)
+                : Dynamic{dynamic}
+            {
+                const bgfx::Memory* memory = bgfx::copy(bytes.As<Napi::Uint8Array>().Data(), bytes.ByteLength());
+
+                if (Dynamic)
+                {
+                    m_handleIndex = bgfx::createDynamicIndexBuffer(memory, flags).idx;
+                }
+                else
+                {
+                    m_handleIndex = bgfx::createIndexBuffer(memory, flags).idx;
+                }
+            }
+
+            ~IndexBufferData()
+            {
+                if (Dynamic)
+                {
+                    bgfx::destroy(bgfx::DynamicIndexBufferHandle{m_handleIndex});
+                }
+                else
+                {
+                    bgfx::destroy(bgfx::IndexBufferHandle{m_handleIndex});
+                }
+            }
+
+            void Update(Napi::TypedArray bytes, uint32_t startingIdx)
+            {
+                const bgfx::Memory* memory = bgfx::copy(bytes.As<Napi::Uint8Array>().Data(), static_cast<uint32_t>(bytes.ByteLength()));
+                bgfx::update(bgfx::DynamicIndexBufferHandle{m_handleIndex}, startingIdx, memory);
+            }
+
+            std::variant<bgfx::IndexBufferHandle, bgfx::DynamicIndexBufferHandle> Handle() const
+            {
+                if (Dynamic)
+                {
+                    return bgfx::DynamicIndexBufferHandle{m_handleIndex};
+                }
+                else
+                {
+                    return bgfx::IndexBufferHandle{m_handleIndex};
+                }
+            }
+
+            const bool Dynamic{};
+
+        private:
+            uint16_t m_handleIndex{bgfx::kInvalidHandle};
+        };
+
+        class VertexBufferData final
```

Comment:
This bool with a variant return seems annoying. Would it be cleaner to use two different classes VertexBufferData and DynamicVertexBufferData and `new` a variant of these?

Reply:
It's a bit clunky, but without this we'd have to pass down a Boolean `isDynamic` parameter (or equivalent) from the JavaScript to the Create and Record methods for both vertex and index buffers; otherwise we won't know what kind of `Napi::External` to to interpret our argument as. I did something like that initially, but it was even clunkier than this and added unnecessary complexity to the JS-native boundary, and the few other ideas I considered were clunkier still, so I ended up using a `std::variant` as the least ugly of many ugly options.

Reply:
I'm not sure you understand what I mean or I'm missing something. If you pass std::variant<VertexBufferData, DynamicVertexBufferData>* as the data to NAPI, the you can tell which one it is when you receive it?